File: packages/autonomo-kotlin/bin/main/com/autonomo/Commands.kt
package com.autonomo

/**
 * Result of a command execution
 */
data class CommandResult(
    val success: Boolean,
    val state: AppState,
    val message: String? = null,
    val error: String? = null
) {
    fun toMap(): Map<String, Any?> = buildMap {
        put("success", success)
        put("state", state.toMap())
        message?.let { put("message", it) }
        error?.let { put("error", it) }
    }

    companion object {
        fun ok(message: String? = null) =
            CommandResult(success = true, state = StateManager.getState(), message = message)

        fun fail(error: String) =
            CommandResult(success = false, state = StateManager.getState(), error = error)
    }
}

/**
 * Navigation handler type
 */
typealias NavigationHandler = (String) -> Unit

/**
 * Command execution
 */
object Commands {
    var navigationHandler: NavigationHandler? = null

    /**
     * Navigate to a screen
     */
    fun navigate(screen: String): CommandResult {
        val handler = navigationHandler ?: return CommandResult.fail("No navigation handler registered")
        handler(screen)
        Thread.sleep(100)
        return CommandResult.ok("Navigated to $screen")
    }

    /**
     * Press/tap an element
     */
    fun press(elementId: String): CommandResult {
        val handler = ElementRegistry.get(elementId)
            ?: return CommandResult.fail("Element not found: $elementId. Available: ${ElementRegistry.list().joinToString(", ")}")
        
        if (handler.disabled) {
            return CommandResult.fail("Element is disabled: $elementId")
        }

        handler.handler(null)
        Thread.sleep(100)
        return CommandResult.ok("Pressed $elementId")
    }

    /**
     * Fill text into an input element
     */
    fun fill(elementId: String, value: String): CommandResult {
        val handler = ElementRegistry.get(elementId)
            ?: return CommandResult.fail("Element not found: $elementId. Available: ${ElementRegistry.list().joinToString(", ")}")

        if (handler.type != ElementType.INPUT) {
            return CommandResult.fail("Element $elementId is not an input (type: ${handler.type.value})")
        }

        if (handler.disabled) {
            return CommandResult.fail("Element is disabled: $elementId")
        }

        handler.handler(value)
        Thread.sleep(50)
        return CommandResult.ok("Filled $elementId with \"$value\"")
    }

    /**
     * Submit an input (press enter)
     */
    fun submit(elementId: String): CommandResult {
        val handler = ElementRegistry.get(elementId)
            ?: return CommandResult.fail("Element not found: $elementId")

        val onSubmit = handler.onSubmit
            ?: return CommandResult.fail("Element $elementId does not support submit")

        onSubmit()
        Thread.sleep(100)
        return CommandResult.ok("Submitted $elementId")
    }

    /**
     * Execute a custom action
     */
    fun custom(actionName: String, value: String? = null): CommandResult {
        val result = CustomActionsRegistry.execute(actionName, value)
        return CommandResult(
            success = result.success,
            state = StateManager.getState(),
            message = result.message,
            error = result.error
        )
    }

    /**
     * Wait for a duration
     */
    fun wait(ms: Int): CommandResult {
        Thread.sleep(ms.toLong())
        return CommandResult.ok("Waited ${ms}ms")
    }

    /**
     * Get current state without any action
     */
    fun getState(): CommandResult = CommandResult.ok()

    /**
     * Execute a command by type
     */
    fun execute(command: String, target: String? = null, value: String? = null): CommandResult =
        when (command.lowercase()) {
            "navigate" -> {
                if (target == null) CommandResult.fail("Navigate requires a target screen")
                else navigate(target)
            }

            "press", "tap", "click" -> {
                if (target == null) CommandResult.fail("Press requires a target element ID")
                else press(target)
            }

            "fill", "type" -> {
                if (target == null) CommandResult.fail("Fill requires a target element ID")
                else fill(target, value ?: "")
            }

            "submit" -> {
                if (target == null) CommandResult.fail("Submit requires a target element ID")
                else submit(target)
            }

            "custom" -> {
                if (target == null) CommandResult.fail("Custom requires an action name")
                else custom(target, value)
            }

            "wait" -> wait(target?.toIntOrNull() ?: 1000)

            "state", "snapshot" -> getState()

            else -> CommandResult.fail("Unknown command: $command")
        }
}

/**
 * Set the navigation handler
 */
fun setNavigationHandler(handler: NavigationHandler) {
    Commands.navigationHandler = handler
}

/**
 * Execute a command
 */
fun executeCommand(command: String, target: String? = null, value: String? = null): CommandResult =
    Commands.execute(command, target, value)

File: packages/Autonomo.CSharp/Commands.cs
namespace Autonomo;

/// <summary>
/// Result of a command execution
/// </summary>
public class CommandResult
{
    public bool Success { get; init; }
    public AppState State { get; init; } = null!;
    public string? Message { get; init; }
    public string? Error { get; init; }

    public Dictionary<string, object> ToDictionary()
    {
        var result = new Dictionary<string, object>
        {
            ["success"] = Success,
            ["state"] = State.ToDictionary()
        };
        if (Message != null) result["message"] = Message;
        if (Error != null) result["error"] = Error;
        return result;
    }

    public static CommandResult Ok(string? message = null)
        => new() { Success = true, Message = message, State = StateManager.Instance.GetState() };

    public static CommandResult Fail(string error)
        => new() { Success = false, Error = error, State = StateManager.Instance.GetState() };
}

/// <summary>
/// Navigation handler delegate
/// </summary>
public delegate void NavigationHandler(string screen);

/// <summary>
/// Command execution service
/// </summary>
public static class Commands
{
    private static NavigationHandler? _navigationHandler;

    /// <summary>
    /// Set the navigation handler
    /// </summary>
    public static void SetNavigationHandler(NavigationHandler handler)
    {
        _navigationHandler = handler;
    }

    /// <summary>
    /// Navigate to a screen
    /// </summary>
    public static CommandResult Navigate(string screen)
    {
        try
        {
            if (_navigationHandler == null)
            {
                return CommandResult.Fail("No navigation handler registered");
            }
            _navigationHandler(screen);
            Thread.Sleep(100);
            return CommandResult.Ok($"Navigated to {screen}");
        }
        catch (Exception ex)
        {
            return CommandResult.Fail(ex.Message);
        }
    }

    /// <summary>
    /// Press/tap an element
    /// </summary>
    public static CommandResult Press(string elementId)
    {
        var handler = ElementRegistry.Instance.Get(elementId);
        if (handler == null)
        {
            var available = string.Join(", ", ElementRegistry.Instance.List());
            return CommandResult.Fail($"Element not found: {elementId}. Available: {available}");
        }
        if (handler.Disabled)
        {
            return CommandResult.Fail($"Element is disabled: {elementId}");
        }
        try
        {
            handler.Handler(null);
            Thread.Sleep(100);
            return CommandResult.Ok($"Pressed {elementId}");
        }
        catch (Exception ex)
        {
            return CommandResult.Fail(ex.Message);
        }
    }

    /// <summary>
    /// Fill text into an input element
    /// </summary>
    public static CommandResult Fill(string elementId, string value)
    {
        var handler = ElementRegistry.Instance.Get(elementId);
        if (handler == null)
        {
            var available = string.Join(", ", ElementRegistry.Instance.List());
            return CommandResult.Fail($"Element not found: {elementId}. Available: {available}");
        }
        if (handler.Type != ElementType.Input)
        {
            return CommandResult.Fail($"Element {elementId} is not an input (type: {handler.Type})");
        }
        if (handler.Disabled)
        {
            return CommandResult.Fail($"Element is disabled: {elementId}");
        }
        try
        {
            handler.Handler(value);
            Thread.Sleep(50);
            return CommandResult.Ok($"Filled {elementId} with \"{value}\"");
        }
        catch (Exception ex)
        {
            return CommandResult.Fail(ex.Message);
        }
    }

    /// <summary>
    /// Submit an input (press enter)
    /// </summary>
    public static CommandResult Submit(string elementId)
    {
        var handler = ElementRegistry.Instance.Get(elementId);
        if (handler == null)
        {
            return CommandResult.Fail($"Element not found: {elementId}");
        }
        if (handler.OnSubmit == null)
        {
            return CommandResult.Fail($"Element {elementId} does not support submit");
        }
        try
        {
            handler.OnSubmit();
            Thread.Sleep(100);
            return CommandResult.Ok($"Submitted {elementId}");
        }
        catch (Exception ex)
        {
            return CommandResult.Fail(ex.Message);
        }
    }

    /// <summary>
    /// Execute a custom action
    /// </summary>
    public static CommandResult Custom(string actionName, string? value = null)
    {
        var result = CustomActionsRegistry.Instance.Execute(actionName, value);
        return new CommandResult
        {
            Success = result.Success,
            Message = result.Message,
            Error = result.Error,
            State = StateManager.Instance.GetState()
        };
    }

    /// <summary>
    /// Wait for a duration
    /// </summary>
    public static CommandResult Wait(int ms)
    {
        Thread.Sleep(ms);
        return CommandResult.Ok($"Waited {ms}ms");
    }

    /// <summary>
    /// Get current state without any action
    /// </summary>
    public static CommandResult GetState()
    {
        return CommandResult.Ok();
    }

    /// <summary>
    /// Execute a command by type
    /// </summary>
    public static CommandResult Execute(string command, string? target = null, string? value = null)
    {
        var cmd = command.ToLower();

        return cmd switch
        {
            "navigate" when target == null =>
                CommandResult.Fail("Navigate requires a target screen"),
            "navigate" =>
                Navigate(target!),

            "press" or "tap" or "click" when target == null =>
                CommandResult.Fail("Press requires a target element ID"),
            "press" or "tap" or "click" =>
                Press(target!),

            "fill" or "type" when target == null =>
                CommandResult.Fail("Fill requires a target element ID"),
            "fill" or "type" =>
                Fill(target!, value ?? ""),

            "submit" when target == null =>
                CommandResult.Fail("Submit requires a target element ID"),
            "submit" =>
                Submit(target!),

            "custom" when target == null =>
                CommandResult.Fail("Custom requires an action name"),
            "custom" =>
                Custom(target!, value),

            "wait" =>
                Wait(int.TryParse(target, out var ms) ? ms : 1000),

            "state" or "snapshot" =>
                GetState(),

            _ => CommandResult.Fail($"Unknown command: {command}")
        };
    }
}
